lxd/instance/drivers/driver_qemu.go

```diff
@@ -7426,22 +7432,91 @@ func (d *qemu) DeviceEventHandler(runConf *deviceConfig.RunConfig) error {
 	return nil
 }
 
-// vsockID returns the vsock Context ID for the VM.
-func (d *qemu) vsockID() int {
-	// We use the system's own VsockID as the base.
-	//
-	// This is either "2" for a physical system or the VM's own id if
-	// running inside of a VM.
-	//
-	// To this we add 1 for backward compatibility with prior logic
-	// which would start at id 3 rather than id 2. Removing that offset
-	// would cause conflicts between existing VMs until they're all rebooted.
-	//
-	// We then add the VM's own instance id (1 or higher) to give us a
-	// unique, non-clashing context ID for our guest.
+// getVsockID returns the vsock Context ID for the VM.
+// If no error is returned, a vsock Context ID of 0 indicates the value is not set for the VM.
+func (d *qemu) getVsockID() (int, error) {
+	existingVsockID, ok := d.localConfig["volatile.vsock_id"]
+	if ok {
+		vsockID, err := strconv.Atoi(existingVsockID)
+		if err != nil {
+			return 0, fmt.Errorf("Failed to parse volatile.vsock_id: %q: %w", existingVsockID, err)
+		}
 
-	info := DriverStatuses()[instancetype.VM].Info
-	return info.Features["vhost_vsock"].(int) + 1 + d.id
+		return vsockID, nil
+	}
+
+	return 0, nil
+}
+
+// freeVsockID returns true if the given vsockID is not yet acquired.
+func (d *qemu) freeVsockID(vsockID uint32) bool {
+	c, err := lxdvsock.Dial(vsockID, shared.HTTPSDefaultPort)
+	if err != nil {
+		var unixErrno unix.Errno
+
+		if !errors.As(err, &unixErrno) {
+			return false
+		}
+
+		if unixErrno == unix.ENODEV {
+			// The syscall to the vsock device returned "no such device".
+			// This means the address (Context ID) is free.
+			return true
+		}
+	}
+
+	// Address is already in use.
+	c.Close()
+	return false
+}
+
+// nextVsockID returns the next free vsock Context ID for the VM.
+// It tries to acquire one randomly until the timeout exceeds.
+func (d *qemu) nextVsockID() (int, error) {
```

Comment:
can we use uint32 throughout these functions rather than sometimes using uint32 and sometimes using int?
Or is there a reason for that?

Reply:
I wanted to keep the impact low but happy to change it here and update all the other references too.

Reply:
Yes, but we can split into separate commits. I suspect the way to go with this is to first change the types to uint32 everywhere.

Then a commit to address the "client" use-case (of getting the instance's vsock ID for purposes of connecting to it's agent).

And then finally a commit to update the "allocation" use-case (of getting a free vsock ID for purposes of starting the instance).

Reply:
Commits are splitted now logically.